### PR TITLE
Sample Django Weather Application

### PR DESCRIPTION
**Pull request by salonijain0918**
Sample application for pybrake on Sanic framework for swagger specification defined in https://confluence.logicmonitor.com/pages/viewpage.action?pageId=156305152

Sample application for weather with below endpoints:

GET /date - returns the date of the server
GET /locations - returns an array of location names
GET /weather/ - returns weather information about the specified location

### CODE REVIEW

**Comment by scottsbaldwin** [comment]
@salonijain0918 what's the difference between this PR and #169?

File: examples/django/django_exampleapp/README.md
Comment (thompiler) [inline examples/django/django_exampleapp/README.md:17]: If the user is expected to replace the default `project_id` and `project_key` values in [`examples/django/django_exampleapp/django_exampleapp/settings.py`](https://github.com/airbrake/pybrake/pull/177/files#diff-543d7436e08bb0cef2501c0c0138906c62bd46db02e01a43e035ecd56055070bR31-R34), then we should add that as a step here right?

File: examples/django/django_exampleapp/weather_app/urls.py
Comment (thompiler) [inline examples/django/django_exampleapp/weather_app/urls.py:10]: I think adding these trailing slashes means that they're required in the curl which is _probably_ not what we want.

Example:
1. Run: `pip3 install -r requirements.txt`
2. Run: `python manage.py runserver 3000`
3. Run: `curl -i "http://localhost:3000/locations"` Get:
```
HTTP/1.1 301 Moved Permanently
Date: Wed, 05 Jan 2022 23:33:31 GMT
Server: WSGIServer/0.2 CPython/3.9.7
Content-Type: text/html; charset=utf-8
Location: /locations/
Content-Length: 0
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin
```
4. Run: `curl -i "http://localhost:3000/locations/"` (note that trailing `/` character) Get:
```
HTTP/1.1 200 OK
Date: Wed, 05 Jan 2022 23:34:46 GMT
Server: WSGIServer/0.2 CPython/3.9.7
Content-Type: application/json
X-Frame-Options: DENY
Content-Length: 34
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin

["pune", "austin", "santabarbara"]%
```

Comment (scottsbaldwin) [reply examples/django/django_exampleapp/weather_app/urls.py:10]: An alternative would be to use `curl -iL ...` where the `-L` will allow the `Location` header from the 301 response. However, I think you're right Thom that we should look at removing the trailing slash.

File: examples/django/django_exampleapp/README.md
Comment (thompiler) [inline examples/django/django_exampleapp/README.md:33]: I personally find these `b761be830f7c23ebe1c3250d42c43673` API keys kind of confusing because they look just like Airbrake project API keys. I could imagine a customer getting confused by this. 

I'm not sure a great way around this. Maybe a value like these would be more obviously unrelated to Airbrake credentials:
- `weather-app-key`
- `weather-example-app-key-123`
- `123weatherAppKey123`

**Comment by thompiler** [comment]
Which out of these commands from the [setup readme ](https://github.com/airbrake/pybrake/blob/386e793d14c8006b2f8fabdff06eba9dd2847449/examples/django/django_exampleapp/README.md)should result in Airbrake errors being sent? 

```
curl "http://localhost:3000/date" -H 'api-key: b761be830f7c23ebe1c3250d42c43673' 
curl "http://localhost:3000/locations" -H 'api-key: b761be830f7c23ebe1c3250d42c43673' 
curl "http://localhost:3000/weather/<austin/pune/santabarbara>" -H 'api-key: b761be830f7c23ebe1c3250d42c43673' 
curl "http://localhost:3000/weather/" -H 'api-key: b761be830f7c23ebe1c3250d42c43673'
```

I was not able to get any Airbrake error to send unless I specified a city which was not supported. These are the things I ran:

#### No error
```
curl -i "http://localhost:3000/date/" -H 'api-key: b761be830f7c23ebe1c3250d42c43673' 
```
#### No error
```
curl -i "http://localhost:3000/locations/" -H 'api-key: b761be830f7c23ebe1c3250d42c43673' 
```
#### No error
```
curl -i "http://localhost:3000/weather/austin" -H 'api-key: b761be830f7c23ebe1c3250d42c43673' 
```
#### Console error but no Airbrake error
```
curl -i "http://localhost:3000/weather/" -H 'api-key: b761be830f7c23ebe1c3250d42c43673'
```
#### Console error and Airbrake error
```
curl -i "http://localhost:3000/weather/tokyo" -H 'api-key: b761be830f7c23ebe1c3250d42c43673' 
```
<details>
<summary>
Click to view screenshot of Airbrake error
</summary>
Screen Shot 2022-01-05 at 5 04 15 PM

</details>

File: examples/django/django_exampleapp/README.md
Comment (scottsbaldwin) [reply examples/django/django_exampleapp/README.md:33]: It's just a hard coded value, and we could change it to something more descriptive like @thompiler suggested.

Comment (thompiler) [reply examples/django/django_exampleapp/README.md:33]: Hey @scottsbaldwin, should this `api-key` being passed in the curl commands for the weather app be authenticated in any way? @salonijain0918 pointed out to me that Chiman's go version of the weather app does have some basic [authentication middleware](https://github.com/airbrake/gobrake/pull/227/files#diff-dc97aeac29b833333a6571446a5b2a501d20d701aad70238f9625b08fc5b3d94R108-R125). The [swagger](https://gist.githubusercontent.com/scottsbaldwin/5f87c2723232f702d5bb434e587c12f2/raw/f6dcbe5208aaf4b230a50445d65faa882ddb326d/weather-api.yaml) for this app though only defines 200 responses (no 401 responses as far as I can see).

Some possible ways the weather app could act:
- Always returns 200 for all curls to valid routes regardless of the value of `api-key` in the header or whether or not it is present.
- Always returns 200s for all curls to valid routes if some `api-key` header is present (ie: it can be any value), otherwise returns 401.
- Only returns 200s for all curls to valid routes if a specific `api-key` header is present (ie: only the api-key in the example would be accepted). (This is what Chiman's example app does), otherwise, return 401.

**Comment by salonijain0918** [comment]
> @salonijain0918 what's the difference between this PR and #169?

Hey Scott, I think I hadn't done a forced update after rebase and my changes couldn't get synced with mainline changes, Thomas and I discussed to continue with this PR for now, will avoid multiple PRs from next time.

File: examples/django/django_exampleapp/README.md
Comment (salonijain0918) [reply examples/django/django_exampleapp/README.md:17]: Hey @thompiler, added the suggested point in README file. I will get a confirmation about your other suggestion as well and update it.

File: examples/django/django_exampleapp/weather_app/urls.py
Comment (salonijain0918) [reply examples/django/django_exampleapp/weather_app/urls.py:10]: I have updated the trailing slashes in the README examples as it is the convention followed for the routes in URL patterns in django, so removing the trailing slash would not work for our application.

File: examples/django/django_exampleapp/README.md
Comment (salonijain0918) [reply examples/django/django_exampleapp/README.md:37]: Hey @scottsbaldwin, Django framework has its own in-built middleware for TokenAuthentication, but it requires the use of database to validate the token, I did some research and had a word with @thompiler as well.
and that would also require use of separate route for validation.

Comment (salonijain0918) [reply examples/django/django_exampleapp/README.md:37]: We can also make use of JWT which would generate a fresh token for each session, please suggest.

Comment (scottsbaldwin) [reply examples/django/django_exampleapp/README.md:37]: I've updated the swagger and removed the `api-key` requirement. Let's not do that part.

**Review by scottsbaldwin — APPROVED** [review]
LGTM

**Comment by scottsbaldwin** [comment]
I approved the PR, but I have one question about the naming convention of the directories. Having multiple directories named `django_exampleapp` in this path seems a bit confusing: `examples/django/django_exampleapp/django_exampleapp/`. Is there a better way to name these directories?

File: examples/django/README.md
Comment (scottsbaldwin) [inline examples/django/README.md:16]: You need to add pybrake to `requirements.txt`. When I do `pip3 install` and then `python manage.py runserver 3000`, I get:

```
ModuleNotFoundError: No module named 'pybrake'
```

File: examples/django/requirements.txt
Comment (scottsbaldwin) [inline examples/django/requirements.txt:14]:
```suggestion
pybrake==1.0.4
python-oauth2==1.1.1
```

File: examples/django/README.md
Comment (scottsbaldwin) [inline examples/django/README.md:34]:
```suggestion
curl "http://localhost:3000/date/"
```

File: examples/django/README.md
Comment (scottsbaldwin) [inline examples/django/README.md:35]:
```suggestion
curl "http://localhost:3000/locations/"
```

File: examples/django/README.md
Comment (scottsbaldwin) [inline examples/django/README.md:36]:
```suggestion
curl "http://localhost:3000/weather/<austin/pune/santabarbara>"
```

File: examples/django/README.md
Comment (scottsbaldwin) [inline examples/django/README.md:37]:
```suggestion
# Should produce an HTTP 404 error
curl -I "http://localhost:3000/weather/"

# Should produce an HTTP 404 error and report the error to Airbrake (since `washington` is not in the supported cities list)
curl -I "http://localhost:3000/weather/washington"
```

File: examples/django/README.md
Comment (scottsbaldwin) [inline examples/django/README.md:40]:
```suggestion

```

**Review by scottsbaldwin — CHANGES_REQUESTED** [review]

File: examples/django/app/views.py
Comment (scottsbaldwin) [inline examples/django/app/views.py:39]: There is a bug here, but let's keep it and explain why it's a bug.
```suggestion
    # This is an intentional bug that occurs when a `location_name` is passed but is not in `city_list`.
    return JsonResponse(data, status=status.HTTP_200_OK, safe=False)
```

The error is:
```
UnboundLocalError:  local variable 'data' referenced before assignment
```

I found this by running the example app and then using this curl command:

```
curl -I "http://localhost:3000/weather/washington"
```

In Airbrake, I received the error explaining the bug.

File: examples/django/README.md
Comment (scottsbaldwin) [inline examples/django/README.md:41]:
```suggestion
# Should produce an intentional HTTP 500 error and report the error to Airbrake (since `washington` is not in the supported cities list, an `if` condition is bypassed and the `data` variable is used but not initialized)
```

**Review by scottsbaldwin — CHANGES_REQUESTED** [review]
Almost there!

File: examples/django/README.md
Comment (salonijain0918) [reply examples/django/README.md]: Hey Scott, the response was being returned outside of the try block, fixed that, now we should be getting proper ValueError instead of UnboundLocalError.

File: examples/django/app/views.py
Comment (scottsbaldwin) [inline examples/django/app/views.py:33]: What happens with `location_name` is not in `city_list`? What response will be sent? The code only returns a response if the `location_name` is supported. What happens with this curl:

```
curl -I "http://localhost:3000/weather/washington"
```

**Review by scottsbaldwin — CHANGES_REQUESTED** [review]

File: examples/django/app/views.py
Comment (salonijain0918) [reply examples/django/app/views.py:33]: The `location_name` is just a vari

Comment (scottsbaldwin) [reply examples/django/app/views.py:33]: If `location_name` is in `city_list`, then there are two possible outcomes:
1. return HTTP 200 with JSON if we have a json file with the same name as the `location_name`
2. return a HTTP 404 ("Not Found") but with status of HTTP 500 - this is not correct to mix status codes

There is NO "else" condition to catch when `location_name` is NOT in `city_list`.

Comment (smurf-U) [reply examples/django/app/views.py]: @scottsbaldwin 

As mentioned in the previous two statements, the situation has improved.

**Comment by smurf-U** [comment]
@kyrylo @scottsbaldwin Could you please complete the final review so that we can close this PR?

**Review by scottsbaldwin — CHANGES_REQUESTED** [review]
I get an error when running the instructions for the Django example.

File: examples/django/README.md
Comment (smurf-U) [reply examples/django/README.md:38]: Updated changed with backward compatible in PR #190.

Comment (kyrylo) [reply examples/django/README.md:38]: Approved #190

**Comment by kyrylo** [comment]
How's the progress here?

**Comment by smurf-U** [comment]
Once #192 will merge, it'll fix everything.

**Comment by kyrylo** [comment]
Why not to ship a fixed example app? I am not sure how #192 can fix this PR.

**Comment by smurf-U** [comment]
> Why not to ship a fixed example app? I am not sure how #192 can fix this PR.

Actually, I did not release pybrake new version after #190 merge.

**Comment by kyrylo** [comment]
That's my point. Why not merge the example app and then ship a new version?

File: examples/django/README.md
Comment (smurf-U) [reply examples/django/README.md:38]: You should now be able to run it.

File: examples/django/README.md
Comment (kyrylo) [inline examples/django/README.md:19]: I am a Django noob, so I had to search for this `settings.py`. Let's change that to `weather/settings.py`.

File: examples/django/README.md
Comment (kyrylo) [inline examples/django/README.md:28]: I don't think we want to advertise these options. I would remove lines 28-31. This should work out of the box already.

File: examples/django/README.md
Comment (kyrylo) [inline examples/django/README.md:54]: `washington` doesn't produce an error 500 for me. I think this has to be looked into.

```
% curl -I "http://localhost:3001/weather/washington"
HTTP/1.1 404 Not Found
Date: Thu, 05 May 2022 09:38:17 GMT
Server: WSGIServer/0.2 CPython/3.9.12
Content-Type: application/json
X-Frame-Options: DENY
Content-Length: 32
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin
```

File: examples/django/weather/urls.py
Comment (kyrylo) [inline examples/django/weather/urls.py:19]:
```suggestion
    path('', include('app.urls')),
```

File: examples/django/README.md
Comment (smurf-U) [reply examples/django/README.md:28]: Even `query_stats` and `queue_stats` cannot be disabled by the user?

Added those option over here based on https://github.com/airbrake/pybrake/blob/27c8906e8fe7e6e36faac3c267dbbf0f8184a15e/src/pybrake/notifier.py#L49

https://github.com/airbrake/pybrake/blob/27c8906e8fe7e6e36faac3c267dbbf0f8184a15e/src/pybrake/notifier.py#L50

File: examples/django/README.md
Comment (smurf-U) [reply examples/django/README.md:53]: Updated, Please check once.

**Comment by smurf-U** [comment]
PTAL

**Pull request merged**
